### DISCUSSION TITLE
Firebase locally

### DISCUSSION THREAD

**Discussion by irux** [Web3Auth/Web3Auth#630] [Help]
Hello, I am trying to log in with firebase locally with the emulator as this [guide](https://web3auth.io/docs/guides/firebase) show you, but I am getting a 404 when I try to use the method `connectTo`

here the response:

```
{
  "success": false,
  "message": "Invalid public key"
}
```

does someone have experience with this ?

**Comment by irux** [comment]
I am looking more deeply to the problem and I think the problem is that the emulator doesn't sign the jwt tokens. That means when I give the jwk endpoint of the guide, it will fail because the tokens are not signed. I can theoretically write an endpoint with empty keys that would accept this token, right ? 

I tried something like this as JWKS https://laprop-test.s3.filebase.com/test.json but it didn't work. Do the library support hs256 ?

**Reply by irux** [reply]
I tried something more simple... this: 


```
const initWeb3Wallet = async () => {


  const clientId = 'MYID';
  const web3auth = new Web3Auth({
    clientId,
    chainConfig: { // this is ethereum chain config, change if other chain(Solana, Polygon)
      chainNamespace: CHAIN_NAMESPACES.EIP155,
      rpcTarget: 'https://mainnet.infura.io/v3/776218ac4734478c90191dde8cae483c',
    },
  });
  await Promise.resolve();
  await web3auth.initModal();
}
```
and I get this:
![image](https://user-images.githubusercontent.com/10131986/191864998-c4c33ee5-307a-4ac0-bb37-58a4e4e224bf.png)

**Comment by rinzler99** [comment]
Hey @irux , I am assuming that you are using custom authentication. For custom auth you should use web3authCore sdk not web3auth sdk. As I can see in your code your are using web3auth sdk. Please refer to below code when you want to use custom auth.

```
import { Web3AuthCore } from "@web3auth/core";
import { CHAIN_NAMESPACES } from "@web3auth/base";

const web3auth = new Web3AuthCore({
  chainConfig: {
    clientId: "YOUR_WEB3AUTH_CLIENT_ID",
    chainNamespace: CHAIN_NAMESPACES.EIP155, // SOLANA, OTHER
    chainId: "0x1",
  },
});
await web3auth.init();
```

**Reply by irux** [reply]
Unfortunately with both I get this https://github.com/Web3Auth/Web3Auth/discussions/630#discussioncomment-3713134

**Reply by rinzler99** [reply]
ohh, can you share your code snippet for web3authcore configuration, as web3authCore requires openlogin adapter too I wanna take a look at it.

**Comment by himanshuchawla009** [comment]
@irux , You can ignore that 404 error, Is it blocking your login flow? The purpose of that call is to validate the user session and to do that we create a random session priv key  on the frontend and send the corresponding pub key to backend to check if session exist, it returns 404 always if session doesn't exists on the first call but it should not block your login flow and also we don't support hs256 keys for jwts

**Reply by irux** [reply]
hello @himanshuchawla009 and @rinzler99,

a couple of updates from my side:

1.) yes, as I expected the tokens from the firebase emulator are not signed and because of that we can't use the jwks endpoint from google. What I am doing right now is mocking the function where I get the token and I sign it on the fly. Now this seems to work, because now the popup screen shows "constructing key" and doesn't give any other error like before. I am using this dummy jwks: https://laprop-test.s3.filebase.com/test.json . When I see the logs in that  popup everything seems normal to me. Right now I can see in my dashboard that I have users active, so it seems to work at least to that point.

2.) Now I have other problem: when the screen closes and I should suppose to begin doing the next operations, I get an error and it seems to block everything.

```ts
  private configAdapters() {
    const openLoginAdapter = new OpenloginAdapter({
      adapterSettings: {
        network: 'testnet',
        uxMode: 'popup',
        clientId: 'myclient',
        loginConfig: {
          jwt: {
            name: 'fake-name',
            verifier: 'laprop-testnet-5',
            typeOfLogin: 'jwt',
            clientId: 'myclient'
          }
        }
      }
    });
    return {
      openLoginAdapter
    };
  };
```

```ts
  private getWeb3AuthCore() {
    const clientId = 'myclient';
    const web3auth = new Web3AuthCore({
      clientId,
      chainConfig: { // this is ethereum chain config, change if other chain(Solana, Polygon)
        chainNamespace: CHAIN_NAMESPACES.EIP155,
        rpcTarget: 'https://data-seed-prebsc-1-s1.binance.org:8545/',
        blockExplorer: 'https://bscscann.com'
      }
    });
    const { openLoginAdapter } = this.configAdapters();
    web3auth.configureAdapter(openLoginAdapter);
    return web3auth;
  }
```

```ts
private async initWeb3Core(token) {
      try {
        await this.web3Auth.init();
        console.log('init');
        const session = await this.web3Auth.connectTo(
          WALLET_ADAPTERS.OPENLOGIN, {
            loginProvider: 'jwt',
            extraLoginOptions: {
              id_token: token
              domain: 'http://localhost:3000',
              verifierIdField: 'sub'
            }
          });
        console.log('printing session');
        console.log(session);
      } catch (e) {
        console.log('having error');
        console.log(e);
        throw e;
      }
  }
```

now when I call `initWeb3Core` (as a fresh setup, that means deleting localstorage, session storage, refreshing token, everything...) I get the window open where It says "constructing the key on 127.0.0.1" and where logs seems to be ok, because if I move quickly before the console lost the object reference I can see how the logs show me my public address and everything. But, when the popup closes, I get inside the catch where it throws the error:
```log
WalletLoginError: Failed to connect with walletFailed to login with openlogin
    at WalletLoginError.fromCode (base.esm.js?e61a:355:1)
    at WalletLoginError.connectionError (base.esm.js?e61a:360:1)
    at OpenloginAdapter.connect (openloginAdapter.esm.js?63c7:143:1)
    at async Web3AuthCore.connectTo (core.esm.js?f233:110:1)
```

after that, without doing a hard clean, I keep getting that error without having a new popup. If I try to use the object I always get this:

```
Uncaught (in promise) WalletLoginError: Wallet is not connectedNo wallet is connected
    at WalletLoginError.fromCode (base.esm.js?e61a:355:1)
    at WalletLoginError.notConnectedError (base.esm.js?e61a:370:1)
    at Proxy.authenticateUser (core.esm.js?f233:129:1)
    at Proxy.eval (WalletProviderType.ts?7f1d:509:1)
    at Generator.next (<anonymous>)
    at fulfilled (WalletProviderType.ts:21:58)
```

I was thinking about creating an own repo where I can reproduce everything but I would like to try to solve this without getting to that point xD 

I hope I was clear enough. Thanks!

**Reply by irux** [reply]
@himanshuchawla009 @rinzler99 after going down the stack while debugging I come to a point where the code wants to execute some middlewares and then it fail because of the postmessage function:

https://github.com/torusresearch/OpenLoginSdk/blob/master/packages/openlogin-jrpc/src/jrpc.ts#L133

```
"Error: Failed to execute 'postMessage' on 'Window': #<Object> could not be cloned.
    at PostMessageStream._postMessage (webpack-internal:///./node_modules/@toruslabs/openlogin-jrpc/dist/openloginJrpc.esm.js:1025:24)
    at PostMessageStream._write (webpack-internal:///./node_modules/@toruslabs/openlogin-jrpc/dist/openloginJrpc.esm.js:167:10)
    at doWrite (webpack-internal:///./node_modules/readable-stream/lib/_stream_writable.js:410:139)
    at writeOrBuffer (webpack-internal:///./node_modules/readable-stream/lib/_stream_writable.js:399:5)
    at Writable.write (webpack-internal:///./node_modules/readable-stream/lib/_stream_writable.js:308:11)
    at ObjectMultiplex.ondata (webpack-internal:///./node_modules/readable-stream/lib/_stream_readable.js:682:20)
    at ObjectMultiplex.emit (webpack-internal:///./node_modules/events/events.js:154:5)
    at addChunk (webpack-internal:///./node_modules/readable-stream/lib/_stream_readable.js:299:12)
    at readableAddChunk (webpack-internal:///./node_modules/readable-stream/lib/_stream_readable.js:281:11)
    at Readable.push (webpack-internal:///./node_modules/readable-stream/lib/_stream_readable.js:242:10)"
```

and some stack:

```
middleware (openloginJrpc.esm.js?4387:363)
(anonymous) (openloginJrpc.esm.js?4387:542)
_runMiddleware (openloginJrpc.esm.js?4387:509)
_runAllMiddleware (openloginJrpc.esm.js?4387:491)
await in _runAllMiddleware (async)
_processRequest (openloginJrpc.esm.js?4387:748)
_handle (openloginJrpc.esm.js?4387:722)
handle (openloginJrpc.esm.js?4387:607)
_rpcRequest (openlogin.esm.js?220a:528)
eval (openlogin.esm.js?220a:918)
_jrpcRequest (openlogin.esm.js?220a:917)
request (openlogin.esm.js?220a:809)
_setPIDData (openlogin.esm.js?220a:935)
request (openlogin.esm.js?220a:820)
_selectedLogin (openlogin.esm.js?220a:709)
login (openlogin.esm.js?220a:694)
connectWithProvider (openloginAdapter.esm.js?63c7:241)
await in connectWithProvider (async)
connect (openloginAdapter.esm.js?63c7:131)
connectTo (core.esm.js?f233:110)
eval (WalletOnBoarding.vue?62ca:89)
eval (index.js??clonedRuleSet-3.use[0]!./node_modules/@quasar/app-webpack/lib/webpack/loader.vue.auto-import-quasar.js??ruleSet[0].use[0]!./node_modules/vue-loader/dist/index.js??ruleSet[0].use[1]!./src/pages/profile-pages/WalletOnBoarding.vue?vue&type=script&lang=ts:18)
__awaiter (index.js??clonedRuleSet-3.use[0]!./node_modules/@quasar/app-webpack/lib/webpack/loader.vue.auto-import-quasar.js??ruleSet[0].use[0]!./node_modules/vue-loader/dist/index.js??ruleSet[0].use[1]!./src/pages/profile-pages/WalletOnBoarding.vue?vue&type=script&lang=ts:14)
debug (WalletOnBoarding.vue?62ca:83)
eval (WalletOnBoarding.vue?610f:89)
eval (runtime-dom.esm-bundler.js?9cec:1481)
callWithErrorHandling (runtime-core.esm-bundler.js?f781:155)
callWithAsyncErrorHandling (runtime-core.esm-bundler.js?f781:164)
invoker (runtime-dom.esm-bundler.js?9cec:369)
```

**Reply by himanshuchawla009** [reply]
can i see your entire code for web3auth integration, if possible can your share a sample github repo to debug.

**Reply by irux** [reply]
Hello @himanshuchawla009 yesterday after more debugging I could find the error. The error happens because of the vuejs reactivity system. If I put the web3authcore instance inside a `ref`, the whole object changes and the serialization of the object fails because of this. Maybe a `JSON.stringify` inside the jrcp calls could help to avoid such problems for others in the future :)

**Reply by irux** [reply]
btw @himanshuchawla009 I am using hs256 for my test environment and it works great! fyi! :)

**Reply by bolisteward** [reply]
hello @irux please can you explain it. Because I'm getting similar error about WalletLoginError: Failed to connect with wallet. Failed to login with openlogin.

**Reply by irux** [reply]
> hello @irux please can you explain it. Because I'm getting similar error about WalletLoginError: Failed to connect with wallet. Failed to login with openlogin.
> 
> 

Hey @bolisteward what is the problem you are having? Are you using Vue framework as me? Do you have some code?

**Reply by masehdh** [reply]
@irux Hey, I'm trying to implement your solution except with flutter and firebase emulator. 
I have a jwks endpoint that is identical to yours.

I also have a custom verifier that points to that jwks endpoint and that checks for the aud claim only. 

I sign the unsigned firebase token with the dart_jsonwebtoken package using HS256, and when I decode it, it is identical in structure to the token I get when I don't use the emulator.  https://pub.dev/packages/dart_jsonwebtoken

But I still get the error that auth shares couldn't be found. Can I ask what exactly your LoginConfigItem looks like, what your extraLoginOptions look like, and what your configuration for your custom verifier is?
<img width="1134" alt="Screenshot 2023-04-12 at 5 27 10 PM" src="https://user-images.githubusercontent.com/60318169/231589065-23068741-dfc5-4c2c-938c-932e8df0cb74.png">
<img width="1134" alt="Screenshot 2023-04-12 at 5 27 26 PM" src="https://user-images.githubusercontent.com/60318169/231589068-27e124ef-3d78-496a-9300-b4471f9aa041.png">
<img width="930" alt="Screenshot 2023-04-12 at 5 28 01 PM" src="https://user-images.githubusercontent.com/60318169/231589175-843c487b-2c19-430e-bfaa-6840e40aaba8.png">
<img width="278" alt="Screenshot 2023-04-12 at 5 28 55 PM" src="https://user-images.githubusercontent.com/60318169/231589348-889558bd-14be-4c98-bba8-5906e6e27593.png">

**Reply by masehdh** [reply]
Ended up figuring it out, for others facing a similar issue. https://community.web3auth.io/t/firebase-local-emulator-does-not-work/3650/4
